utils/python/CIME/buildlib.py

```diff
@@ -0,0 +1,49 @@
+from CIME.XML.standard_module_setup import *
+from CIME.case import Case
+"""
+buildlibx - given the caseroot and component name(cmp), build the stub component
+"""
+
+class buildlib(object):
+    def __init__(self, case, model, comp):
+        self.case = case
+        self.model = model
+        self.comp = comp
+        self.caseroot = case.get_value("CASEROOT")
+        self.casetools = case.get_value("CASETOOLS")
+        self.cimeroot = case.get_value("CIMEROOT")
+        self.mach = case.get_value("MACH")
+        self.objroot = case.get_value("OBJROOT")
+        self.libroot = case.get_value("LIBROOT")
+        self.gmake_j = case.get_value("GMAKE_J")
+        self.gmake = case.get_value("GMAKE")
+
+class buildxlib(buildlib):
+    def buildlib(self):
+        try:
+            out = open('Filepath','w')
+            msg = None
+        except IOError, e:
+            msg = str(e)
+
+        expect(msg is None, msg)
+
+        # Write directory list
+        out.write(os.path.join(self.caseroot, "SourceMods", "src.%s", self.comp)
+                  + "\n")
+        out.write(os.path.join(self.cimeroot, "components", "xcpl_comps",
+                               "xshare") + "\n")
+        out.write(os.path.join(self.cimeroot, "components", "xcpl_comps",
+                               self.comp, "cpl") + "\n")
+        out.close()
+
+        # generate macro values
+        complib = os.path.join(self.libroot, "lib%s.a" % self.model)
+        makefile = os.path.join(self.casetools, "Makefile")
+        macfile = os.path.join(self.caseroot, "Macros.%s" % self.mach)
+
+        # build 
+        stat = run_cmd("%s complib -j %d MODEL=%s COMPLIB=%s -f %s MACFILE=%s"
```

Comment:
Hi Jason,

I am going to post this on slack as well. I don't think we need a classes
structure for buildlib - but rather just a module with reused routines.
I also worked on build.py to address issues that I mentioned in the PR. Not
sure how to coordinate all of this. I am in a NSF site review meeting all
day today.

Mariana

On Tue, Jun 7, 2016 at 6:23 AM, sarich <notifications@github.com> wrote:

> In utils/python/CIME/buildlib.py
> <https://github.com/ESMCI/cime/pull/206#discussion_r66059968>:
>
> > +        # Write directory list
> > +        out.write(os.path.join(self.caseroot, "SourceMods", "src.%s", self.comp)
> > +                  + "\n")
> > +        out.write(os.path.join(self.cimeroot, "components", "xcpl_comps",
> > +                               "xshare") + "\n")
> > +        out.write(os.path.join(self.cimeroot, "components", "xcpl_comps",
> > +                               self.comp, "cpl") + "\n")
> > +        out.close()
> > +
> > +        # generate macro values
> > +        complib = os.path.join(self.libroot, "lib%s.a" % self.model)
> > +        makefile = os.path.join(self.casetools, "Makefile")
> > +        macfile = os.path.join(self.caseroot, "Macros.%s" % self.mach)
> > +
> > +        # build
> > +        stat = run_cmd("%s complib -j %d MODEL=%s COMPLIB=%s -f %s MACFILE=%s"
>
> Thanks for the feedback everyone, I was travelling yesterday and trying to
> catch up now. I'll make the suggested changes, they all make sense -
> documentation fixes - class names change buildlib, buildxlib, buildnml,
> buildxnml to BuildLib, BuildXLib, buildNML, buildXNML - open files using
> 'with open as' - don't do any work in __init__ - fix run_cmd to ignore
> return value (it is not okay to fail) I'm keeping these on the table for
> further discussion without changing anything: - Looking at this some more I
> think we need to change build.py to have build_model to remove the
> following statement: bldroot = exeroot #What is this for? - I also hate
> having identical or almost identical code in multiple file locations.
> Complicates development unnecessarily. Maybe we can fix this with symlinks
> or replace the python with a one-line bash script - I am wondering if we
> should be using OBJROOT and LIBROOT as xml variables. Frankly, OBJROOT only
> makes sense if you are not using shared libraries (which is the case for
> all CESM components now other than clm4.5). It seems that we want to pass
> buildroot to build.py and simply use that without querying case.get_value
> for OBJROOT orand calculate it correctly in here (rather than inside the
> component's buildlib. - - Is the file's name really "Filepath"? (Yes) - In
> that case, buildlib.py would only be used for CIME components - is that the
> consensus? Jason
> … <#m_-1198912587222859129_>
> On Mon, Jun 6, 2016 at 6:55 PM, mvertens ***@***.***> wrote: In
> utils/python/CIME/buildlib.py <#206 (comment)
> <https://github.com/ESMCI/cime/pull/206#discussion_r65990087>>: > + #
> Write directory list > + out.write(os.path.join(self.caseroot,
> "SourceMods", "src.%s", self.comp) > + + "\n") > +
> out.write(os.path.join(self.cimeroot, "components", "xcpl_comps", > +
> "xshare") + "\n") > + out.write(os.path.join(self.cimeroot, "components",
> "xcpl_comps", > + self.comp, "cpl") + "\n") > + out.close() > + > + #
> generate macro values > + complib = os.path.join(self.libroot, "lib%s.a" %
> self.model) > + makefile = os.path.join(self.casetools, "Makefile") > +
> macfile = os.path.join(self.caseroot, "Macros.%s" % self.mach) > + > + #
> build > + stat = run_cmd("%s complib -j %d MODEL=%s COMPLIB=%s -f %s
> MACFILE=%s" Actually, looking at this some more I think we need to change
> build.py to have build_model to remove the following statement: bldroot =
> exeroot #What is this for? and calculate it correctly in here (rather than
> inside the component's buildlib. — You are receiving this because you were
> mentioned. Reply to this email directly, view it on GitHub <
> https://github.com/ESMCI/cime/pull/206/files/9a9c0a054a5a717ec97762ff9fb498dcf958cbe0#r65990087>,
> or mute the thread <
> https://github.com/notifications/unsubscribe/AEs_us-o1LXE5y2NzO_MUhRboOGU9uyAks5qJLNVgaJpZM4IvChx>
> .
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMCI/cime/pull/206/files/9a9c0a054a5a717ec97762ff9fb498dcf958cbe0#r66059968>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHlxE5XL1oClH5XT0DLk-SuPfLp_bNWUks5qJWK3gaJpZM4IvChx>
> .
>